### PR TITLE
Testing

### PR DESCRIPTION
**Pull request by barakadanny**
In this PR, the following are the things achieved:

- Write integration tests with Capybara gem.
- Secure app from n+1 problems.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -55,6 +55,8 @@ group :development, :test do
   gem 'debug', platforms: %i[mri mingw x64_mingw]
 
   gem 'rspec-rails', '>= 3.9.0'
+
+  gem 'database_cleaner'
```
Comment (mikethreels) [inline Gemfile:59]: - [x] capybara should be added to your gemfile:
```
group :development, :test do
   ...
   **gem 'capybara'**
   ...
 end
```

File: spec/views/users/users_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    expect(page).to have_content(@user1)
+  end
+
+  it 'shows number of user posts ' do
+    expect(page).to have_content(@user.post_counter)
+  end
+
+  it 'it should render the view all user posts link' do
+    expect(page).to have_link('See all posts')
+  end
```
Comment (mikethreels) [inline spec/views/users/users_show_spec.rb:39]: - [x] for the user show spec your missing the below tests:
Tests that asserts you can see the user's first 3 posts.
Tests that redirects to the post's show page when you click on a user's post.
Tests that redirects to the user's post's index page when you click on a see all posts.

File: spec/views/posts/post_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  scenario 'display the first comment on a post' do
+    expect(page).to have_content('Hello')
+  end
+
+  scenario 'display the how many comments' do
+    expect(page).to have_content('5')
+  end
+
+  scenario 'display the how many Likes' do
+    expect(page).to have_content('0')
```
Comment (mikethreels) [inline spec/views/posts/post_index_spec.rb]: - [x] please make sure to write tests for all the below senario's:
- Tests that asserts you can see the user's profile picture.
- Tests that asserts you can see the user's username.
- Tests that asserts you can see the number of posts the user has written.
- Tests that asserts you can see a post's title.
- Tests that asserts you can see some of the post's body.
- Tests that asserts you can see the first comments on a post.
- Tests that asserts you can see how many comments a post has.
- Tests that asserts you can see how many likes a post has.
- Tests that asserts you can see a section for pagination if there are more posts than fit on the view.
- Tests that redirects to the post's show page when you click on a post.

File: spec/views/posts/post_show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+  scenario 'disolays number of comments' do
+    expect(page).to have_content('5')
+  end
+
+  scenario 'displays number of likes' do
+    expect(page).to have_content('0')
+  end
+
+  scenario 'displays the content of the post' do
```
Comment (mikethreels) [inline spec/views/posts/post_show_spec.rb:27]: - [ ] please make sure to write tests for all the below senario's:
- Tests that asserts you can see the post's title.
- Tests that asserts you can see who wrote the post.
- Tests that asserts you can see how many comments it has.
- Tests that asserts you can see how many likes it has.
- Tests that asserts you can see the post body.
- Tests that asserts you can see the username of each commentor.
- Tests that asserts you can see the comment each commentor left.

**Review by mikethreels — CHANGES_REQUESTED** [review]
Hi Danny, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- linters are passing

## Required Changes ♻️
- [ ] in order to solve the n+1 problem your post index action should look something like `@user.posts.includes(:comments)` instead of `@posts = @user.posts`

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/views/posts/post_show_spec.rb (hunk 4 above)
Comment (barakadanny) [reply spec/views/posts/post_show_spec.rb:27]: @mikethreels 
Thank you for the review
Note: some of the tests have been made `(Post_show_spect)`, 
I made the requested changes

File: Gemfile (hunk 1 above)
Comment (rloterh) [reply Gemfile:59]: well resolved! ✔️

**Comment by rloterh** [comment]
Hi @barakadanny 
Linter checks are not showing. Kindly ensure to fix it asap!

**Comment by barakadanny** [comment]
@rloterh 
Fixed, 
Kindly check now

File: spec/views/users/users_show_spec.rb (hunk 2 above)
Comment (rloterh) [reply spec/views/users/users_show_spec.rb:39]: well resolved! ✔️

File: spec/views/posts/post_index_spec.rb (hunk 3 above)
Comment (rloterh) [reply spec/views/posts/post_index_spec.rb]: well resolved! ✔️

File: spec/views/posts/post_show_spec.rb (hunk 4 above)
Comment (rloterh) [reply spec/views/posts/post_show_spec.rb:27]: well resolved! ✔️

**Review by rloterh — APPROVED** [review]
## Status: Approved :heavy_check_mark:

Hi Danny,  

Your project is complete! 🏁 Nice job! There is nothing else to say other than... it's time to merge it 🍾 Congratulations!  🎉💯🌟 

## Highlights 👍
- Correctly implemented required integration specs
- `N + 1` problem properly handled

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._

**You can also consider:**
- MIT license link in your ReadMe is broken, kindly endeavor to fix it.

Keep soaring higher! 💻 🍷 🚀 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**